sources/host-ctr/cmd/host-ctr/main.go

```diff
@@ -173,65 +188,73 @@ func runCtr(containerdSocket string, namespace string, containerID string, sourc
-	cgroupPath, err := cgroups.GetOwnCgroup("name=systemd")
+	// Check if the target container already exists. If it does, take over the helm to manage it.
+	container, err := client.LoadContainer(ctx, containerID)
 	if err != nil {
-		log.G(ctx).WithError(err).Error("failed to discover systemd cgroup path")
-		return err
+		if errdefs.IsNotFound(err) {
+			log.G(ctx).WithField("ctr-id", containerID).Info("Container does not exist, proceeding to create it")
+		} else {
+			log.G(ctx).WithField("ctr-id", containerID).WithError(err).Error("failed to retrieve list of containers")
+			return err
+		}
 	}
 
-	// Set up the container spec. See `withSuperpowered` for conditional options
-	// set when configured as superpowered.
-	ctrOpts := containerd.WithNewSpec(
-		oci.WithImageConfig(img),
-		oci.WithHostNamespace(runtimespec.NetworkNamespace),
-		oci.WithHostHostsFile,
-		oci.WithHostResolvconf,
-		// Launch the container under the systemd unit's cgroup
-		oci.WithCgroup(cgroupPath),
-		// Mount in the API socket for the Bottlerocket API server, and the API
-		// client used to interact with it
-		oci.WithMounts([]runtimespec.Mount{
-			{
-				Options:     []string{"bind", "rw"},
-				Destination: "/run/api.sock",
-				Source:      "/run/api.sock",
-			},
-			// Mount in the apiclient to make API calls to the Bottlerocket API server
-			{
-				Options:     []string{"bind", "ro"},
-				Destination: "/usr/local/bin/apiclient",
-				Source:      "/usr/bin/apiclient",
-			},
-			// Mount in the persistent storage location for this container
-			{
-				Options:     []string{"rbind", "rw"},
-				Destination: "/.bottlerocket/host-containers/" + containerID,
-				Source:      "/local/host-containers/" + containerID,
-			}}),
-		// Mount the rootfs with an SELinux label that makes it writable
-		withMountLabel("system_u:object_r:state_t:s0"),
-		// Include conditional options for superpowered containers.
-		withSuperpowered(superpowered),
-	)
+	// If the container doesn't already exist, create it
+	if container == nil {
+		// Get the cgroup path of the systemd service
+		cgroupPath, err := cgroups.GetOwnCgroup("name=systemd")
+		if err != nil {
+			log.G(ctx).WithError(err).Error("failed to discover systemd cgroup path")
+			return err
+		}
 
-	// Create and start the container.
-	container, err := client.NewContainer(
-		ctx,
-		containerID,
-		containerd.WithImage(img),
-		containerd.WithNewSnapshot(containerID+"-snapshot", img),
-		ctrOpts,
-	)
-	if err != nil {
-		log.G(ctx).WithError(err).WithField("img", img.Name).Error("failed to create container")
-		return err
+		// Set up the container spec. See `withSuperpowered` for conditional options
+		// set when configured as superpowered.
+		ctrOpts := containerd.WithNewSpec(
+			oci.WithImageConfig(img),
+			oci.WithHostNamespace(runtimespec.NetworkNamespace),
+			oci.WithHostHostsFile,
+			oci.WithHostResolvconf,
+			// Launch the container under the systemd unit's cgroup
+			oci.WithCgroup(cgroupPath),
```

Comment:
Was this the reason behind `KillMode=mixed` in the unit file - so they'd share a cgroup and systemd could clean them up together?

Reply:
Ah, you're right. I had forgotten. It's so that the host container task will get cleaned up if `host-containers@.service` gets stopped by systemd.

I'll go ahead drop that commit that removed `KillMode=mixed`

Reply:
That doesn't necessarily mean the container resources are cleaned up though, so we still the need the host-ctr mechanism.

Reply:
It seems like the cgroup piece isn't working as intended, but that's not a new problem.

We can fix that and clean up `KillMode=mixed` again at the same time. My concern is really that we'd "fix" this code so the tasks landed in the `host-containers@...` cgroup, and then a service restart would cause them to be killed again.